### PR TITLE
DDPB-2463 Page timeout on user edit page

### PR DESCRIPTION
**Pull request by aiesoftware**
**DO NOT MERGE**
**Addresses:**
* Poor performance for pages with high data volumes due to Doctrine hydration and JMS Serialisation
* Gigantic payloads being sent to frontend Client
* Complex nature of JMS serialisation groups in attempt to reduce payloads, but they are confusing, lack consistency and are an attempt to fix the issue of how can we serialise our entities and all of their complex relationships.

**Proposed solutions**
* Create clear boundaries between what the API knows and what the Client _needs_ to know, using Data Transfer Objects (DTO's) to represent trimmed down versions of our Entities to communicate across the wire
* Avoid unnecessary Doctrine hydration and entity serialisation by using Repositories to write native SQL, and using Assemblers to create DTO's from SQL results. Use Controllers to ask a Repository for a DTO and return a json encoded DTO in the response

**Notes**
* There is still a performance hit on the Client side as it currently still hydrates the response into full blown Entities. We should work towards using DTO's on both sides to further secure the boundary between the Application layer and the Domain layer
* New code in the API has been put in a `v2` folder and namespace, and new endpoint is prefixed with `/v2`

**Result**
* On local dev environment, it took `admin/edit-user` for a user with 950 clients to display from ~37 secs. to ~7 secs. Reduces down to 3 secs. when using DTO on Client side (experimental)

**PHPUnit output**
```
AppBundle\v2\Assembler\ClientAssembler
 [x] assembleFromArray returns DTO from given array
 [x] assembleFromArray throws exception if given insufficient data

AppBundle\v2\Assembler\DeputyAssembler
 [x] assembleFromArray returns DTO from given array
 [x] assembleFromArraythrows exception if given insufficient data

AppBundle\v2\Repository\ClientRepository
 [x] getDtoDataArrayByDeputy returns client data for each client belonging to deputy

AppBundle\v2\Repository\DeputyRepository
 [x] getDtoById returns deputy DTO for given deputy
```

### CODE REVIEW

File: app/config/services/repositories.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  UserRepository:
+    class: AppBundle\Entity\User
+    factory: ["@doctrine.orm.entity_manager", getRepository]
+    arguments:
+      - AppBundle\Entity\User
+
+  AppBundle\v2\Repository\DeputyRepository:
+    class: AppBundle\v2\Repository\DeputyRepository
+    arguments:
+      - '@doctrine.orm.entity_manager'
```
Comment (slizzio) [inline app/config/services/repositories.yml]: If the Repository extends AbstractEntityRepository then it has the entity manager already injected via $this->_em

File: app/config/services/repositories.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      - '@UserRepository'
+      - '@AppBundle\v2\Repository\ClientRepository'
```
Comment (slizzio) [inline app/config/services/repositories.yml]: comment only: Not convinced we need to inject this. AbstractEntityRepository supplies methods to get other repositories but appreciate the need to inject to allow easier unit testing so happy to leave it in. Just conscious that as the repositories grow in complexity you may find developers adding several other repositories to the constructor as they see fit for their needs

File: src/AppBundle/Entity/User.php [hunk 3]
```diff
@@ -12,7 +12,7 @@
  * Users.
  *
  * @ORM\Table(name="dd_user", indexes={@ORM\Index(name="deputy_no_idx", columns={"deputy_no"})})
- * @ORM\Entity(repositoryClass="AppBundle\Entity\Repository\UserRepository")
```
Comment (slizzio) [inline src/AppBundle/Entity/User.php]: why is this removed? (curiosity)

File: app/config/services/repositories.yml (hunk 1 above)
Comment (aiesoftware) [reply app/config/services/repositories.yml]: Technically yes, but composition should be favoured over inheritance. Extending a class (which is already an extension) just to get at a property within a parent is a code smell and probably means we should inject the service we need.

File: src/AppBundle/v2/Assembler/ClientAssembler.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+class ClientAssembler
+{
+    /**
+     * @param array $data
+     * @return ClientDto
+     */
+    public function assembleFromArray(array $data)
+    {
+        $this->throwExceptionIfMissingRequiredData($data);
```
Comment (slizzio) [inline src/AppBundle/v2/Assembler/ClientAssembler.php]: Maybe 
if (!$this->dataIsValid($data)) {           
 throw new \InvalidArgumentException(__CLASS__ . ': Missing all data required to build DTO');
}
or even rename dataIsValid() to validateData() and put it in an interface? to ensure all future assemblers have that method?

File: app/config/services/repositories.yml (hunk 2 above)
Comment (aiesoftware) [reply app/config/services/repositories.yml]: Unit testing is a very valid reason in my opinion. If this repository happens to need another repository, then I think it would be a valid reason to inject one.

File: src/AppBundle/v2/Assembler/ClientAssembler.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        return
+            array_key_exists('id', $data) &&
+            array_key_exists('case_number', $data) &&
+            array_key_exists('firstname', $data) &&
+            array_key_exists('lastname', $data) &&
+            array_key_exists('email', $data) &&
+            array_key_exists('report_count', $data) &&
+            array_key_exists('ndr_id', $data);
+    }
```
Comment (slizzio) [inline src/AppBundle/v2/Assembler/ClientAssembler.php]: (optional) consider using 
   return !empty(array_diff(['id', 'case_number' ... ,'ndr_id'], $data));

File: src/AppBundle/Entity/User.php (hunk 3 above)
Comment (aiesoftware) [reply src/AppBundle/Entity/User.php]: One of the benefits of making the Repository classes standalone, without extending Doctrine, is that we can have many Repo's for one Entity, and therefore don't need to declare this here.

File: src/AppBundle/v2/Assembler/ClientAssembler.php (hunk 5 above)
Comment (slizzio) [reply src/AppBundle/v2/Assembler/ClientAssembler.php]: (optional) consider moving this to an Abstract class that references self::dtoKeys set as a property ?

File: src/AppBundle/v2/Assembler/ClientAssembler.php (hunk 4 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/ClientAssembler.php]: Using the software rule of 3, I'd rather not break anything out into abstract classes or interface methods yet. Your top suggestion is pretty much what is happening inside that method.

File: src/AppBundle/v2/Assembler/DeputyAssembler.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            array_key_exists('address_postcode', $data) &&
+            array_key_exists('odr_enabled', $data) &&
+            array_key_exists('clients', $data);
+    }
+
+    /**
+     * @param $deputy
+     * @return DeputyAssembler
+     */
+    private function buildDeputyDtoFromArray($deputy)
```
Comment (slizzio) [inline src/AppBundle/v2/Assembler/DeputyAssembler.php]: if we lose the 'Deputy' from this method it will allow much easier refactoring later on

Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/DeputyAssembler.php]: Okay

File: src/AppBundle/v2/Assembler/DeputyAssembler.php [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        );
+
+        return $this;
+    }
+
+    /**
+     * @param array $clients
+     * @return DeputyAssembler
+     */
+    private function buildAndAttachClientDtosFromArray(array $clients)
```
Comment (slizzio) [inline src/AppBundle/v2/Assembler/DeputyAssembler.php]: same here. Should just be 'buildDTO()'

File: src/AppBundle/v2/Assembler/ClientAssembler.php (hunk 5 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/ClientAssembler.php]: Would this care about the order of keys? The method doesn't care about order, just existence..

File: src/AppBundle/v2/Assembler/DeputyAssembler.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        $this->deputyDto->setClients($clients);
+
+        return $this;
+    }
+
+    /**
+     * @return DeputyDto
+     */
+    private function getDeputyDto()
```
Comment (slizzio) [inline src/AppBundle/v2/Assembler/DeputyAssembler.php]: getDto()

File: src/AppBundle/v2/Assembler/DeputyAssembler.php (hunk 7 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/DeputyAssembler.php]: This method name tells a story, it helps devs understand what is happening without having to read inside the method.

File: src/AppBundle/v2/Assembler/DeputyAssembler.php (hunk 8 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/DeputyAssembler.php]: Okay.

File: src/AppBundle/v2/Controller/DeputyController.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+     * @Route("/{id}", requirements={"id":"\d+"})
+     * @Method({"GET"})
+     *
+     * @param $id
+     * @return JsonResponse
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getByIdAction($id)
+    {
+        return $this->buildSuccessResponse($this->repository->getDtoById($id));
```
Comment (slizzio) [inline src/AppBundle/v2/Controller/DeputyController.php]: This assumes success

File: src/AppBundle/v2/Assembler/ClientAssembler.php (hunk 5 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Assembler/ClientAssembler.php]: Not keen on forcing the abstraction yet.

File: src/AppBundle/v2/Controller/DeputyController.php (hunk 9 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Controller/DeputyController.php]: I'll rephrase or reorder it 👍

File: src/AppBundle/v2/Controller/DeputyController.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * @param DeputyDto $dto
+     * @return JsonResponse
+     */
+    private function buildSuccessResponse(DeputyDto $dto)
+    {
+        return new JsonResponse([
+            'success' => true,
+            'data' => $dto->jsonSerialize(),
```
Comment (slizzio) [inline src/AppBundle/v2/Controller/DeputyController.php]: Could the Dto not serialize the data before returning?

File: src/AppBundle/v2/DTO/ClientDto.php [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @param $id
+     * @param $caseNumber
+     * @param $firstName
+     * @param $lastName
+     * @param $email
+     * @param $reportCount
+     * @param $ndrId
+     */
+    public function __construct($id, $caseNumber, $firstName, $lastName, $email, $reportCount, $ndrId)
```
Comment (slizzio) [inline src/AppBundle/v2/DTO/ClientDto.php]: would it be better to have the last two parameters optional?

File: src/AppBundle/v2/Controller/DeputyController.php (hunk 10 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Controller/DeputyController.php]: A DTO is an object which will need to be passed through the application as a DTO, not a serialized representation. Make sense?

File: src/AppBundle/v2/DTO/ClientDto.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+     */
+    public function jsonSerialize()
+    {
+        return [
+            'id' => $this->id,
+            'case_number' => $this->caseNumber,
+            'firstname' => $this->firstName,
+            'lastname' => $this->lastName,
+            'email' => $this->email,
+            'total_report_count' => $this->reportCount,
```
Comment (slizzio) [inline src/AppBundle/v2/DTO/ClientDto.php]: id leave as just report_count

Comment (aiesoftware) [reply src/AppBundle/v2/DTO/ClientDto.php]: Needs to match the entity on the client side unfortunately

File: src/AppBundle/v2/DTO/ClientDto.php (hunk 11 above)
Comment (aiesoftware) [reply src/AppBundle/v2/DTO/ClientDto.php]: Maybe, but not sure. Leave it for now?

File: src/AppBundle/v2/DTO/DeputyDto.php [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    public function jsonSerialize()
+    {
+        return [
+            'id' => $this->id,
+            'firstname' => $this->firstName,
+            'lastname' => $this->lastName,
+            'email' => $this->email,
+            'role_name' => $this->roleName,
+            'address_postcode' => $this->postcode,
```
Comment (slizzio) [inline src/AppBundle/v2/DTO/DeputyDto.php]: again, if we keep consistent naming conventions will make it easier to refactor:
address_postcode => $this->addressPostcode

Comment (aiesoftware) [reply src/AppBundle/v2/DTO/DeputyDto.php]: See above

File: src/AppBundle/v2/DTO/DeputyDto.php [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    /**
+     * @return array
+     */
+    private function serializeClients()
+    {
+        $serializedClients = [];
+
+        foreach ($this->clients as $client) {
```
Comment (slizzio) [inline src/AppBundle/v2/DTO/DeputyDto.php]: what happens if clients is null?

File: src/AppBundle/v2/DTO/DeputyDto.php [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        return $this->firstName;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getLastName()
+    {
+        return $this->lastName;
```
Comment (slizzio) [inline src/AppBundle/v2/DTO/DeputyDto.php]: check database fields I think its lastname and firstname, no capital N

File: src/AppBundle/v2/DTO/DeputyDto.php (hunk 14 above)
Comment (aiesoftware) [reply src/AppBundle/v2/DTO/DeputyDto.php]: Good spot, I'll verify it

File: src/AppBundle/v2/Repository/ClientRepository.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        $this->entityManager = $entityManager;
+        $this->clientRepository = $clientRepository;
+    }
+
+    /**
+     * @param $deputyId
+     * @return array
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getDtoDataArrayByDeputy($deputyId)
```
Comment (slizzio) [inline src/AppBundle/v2/Repository/ClientRepository.php]: getDtoData()  then we can put in an interface

File: src/AppBundle/v2/DTO/DeputyDto.php (hunk 15 above)
Comment (aiesoftware) [reply src/AppBundle/v2/DTO/DeputyDto.php]: Do we care about it matching database?

File: src/AppBundle/v2/Repository/DeputyRepository.php [hunk 17]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace AppBundle\v2\Repository;
+
+use AppBundle\v2\Assembler\DeputyAssembler;
+use Doctrine\Common\Persistence\ObjectRepository;
+use Doctrine\ORM\EntityManagerInterface;
+
+class DeputyRepository
```
Comment (slizzio) [inline src/AppBundle/v2/Repository/DeputyRepository.php]: extends AbstractRepository

File: src/AppBundle/v2/Repository/DeputyRepository.php [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+{
+    /** @var EntityManagerInterface */
+    private $entityManager;
```
Comment (slizzio) [inline src/AppBundle/v2/Repository/DeputyRepository.php]: not needed use $this->_em

File: src/AppBundle/v2/Repository/DeputyRepository.php (hunk 17 above)
Comment (aiesoftware) [reply src/AppBundle/v2/Repository/DeputyRepository.php]: See previous comments on that
